src/hotspot/share/code/nmethod.hpp

```diff
@@ -281,23 +289,89 @@ class nmethod : public CompiledMethod {
   address stub_end              () const          { return           header_begin() + _oops_offset          ; }
   address exception_begin       () const          { return           header_begin() + _exception_offset     ; }
   address unwind_handler_begin  () const          { return _unwind_handler_offset != -1 ? (header_begin() + _unwind_handler_offset) : NULL; }
+  // mutable data
   oop*    oops_begin            () const          { return (oop*)   (header_begin() + _oops_offset)         ; }
   oop*    oops_end              () const          { return (oop*)   (header_begin() + _metadata_offset)     ; }
 
   Metadata** metadata_begin   () const            { return (Metadata**)  (header_begin() + _metadata_offset)     ; }
-  Metadata** metadata_end     () const            { return (Metadata**)  _scopes_data_begin; }
-
-  address scopes_data_end       () const          { return           header_begin() + _scopes_pcs_offset    ; }
-  PcDesc* scopes_pcs_begin      () const          { return (PcDesc*)(header_begin() + _scopes_pcs_offset   ); }
-  PcDesc* scopes_pcs_end        () const          { return (PcDesc*)(header_begin() + _dependencies_offset) ; }
-  address dependencies_begin    () const          { return           header_begin() + _dependencies_offset  ; }
-  address dependencies_end      () const          { return           header_begin() + _handler_table_offset ; }
-  address handler_table_begin   () const          { return           header_begin() + _handler_table_offset ; }
-  address handler_table_end     () const          { return           header_begin() + _nul_chk_table_offset ; }
-  address nul_chk_table_begin   () const          { return           header_begin() + _nul_chk_table_offset ; }
-  address nul_chk_table_end     () const          { return           header_begin() + _nmethod_end_offset   ; }
+  Metadata** metadata_end     () const            {
+    if (_immutable_data_size > 0) {
+      return (Metadata**) data_end();
+    } else {
+      return (Metadata**) _scopes_data_begin;
+    }
+  }
+
+  // immutable data
+  address immutable_data_begin() const { return _immutable_data; }
+  address immutable_data_end  () const { return _immutable_data + _immutable_data_size; }
+
+  address scopes_data_end() const {
+    if (_immutable_data_size > 0) {
+      return immutable_data_begin() + _scopes_pcs_offset;
+    } else {
+      return header_begin() + _scopes_pcs_offset;
+    }
+  }
+  PcDesc* scopes_pcs_begin() const {
+    if (_immutable_data_size > 0) {
+      return (PcDesc *) (immutable_data_begin() + _scopes_pcs_offset);
+    } else {
+      return (PcDesc *) (header_begin() + _scopes_pcs_offset);
+    }
+  }
+  PcDesc* scopes_pcs_end() const {
+    if (_immutable_data_size > 0) {
+      return (PcDesc *) (immutable_data_begin() + _dependencies_offset);
+    } else {
+      return (PcDesc *) (header_begin() + _dependencies_offset);
+    }
+  }
```

Comment:
Same with the above comment.
return (PcDesc *) dependencies_begin();

And also some of the following codes could be reusable.